sss/add_puma.py

```diff
@@ -0,0 +1,174 @@
+        puma = puma_txt[puma_txt['sl'] == '795']
+        puma =puma[['puma_code','area_name']]
+        puma = puma.drop_duplicates()
+        puma.columns=['puma_code','puma_area']
+        # left join county and puma by puma code
+        crosswalk = county.merge(puma, how='left')
+        # only keep columns of interests
+        crosswalk = crosswalk[['sl', 'state_fips', 'state','puma_code','county_fips', 'county', 'puma_area','population']]
+        #creat place(sssplace)
+        crosswalk['place'] = crosswalk['county']
+    # attach population of each puma to the crosswalk file
+    crosswalk = crosswalk.merge(puma_pop, on = 'puma_code',suffixes=('_self','_max'))
+    # accoridng to population of each row and population of puma to create a population weight
+    crosswalk['weight'] = crosswalk['population_self']/crosswalk['population_max']
+
+    # TODO: if the puma name and city file is better designed, the code could be easier
+    #if the file is State of Washington, replace some 'place' name to meaningful names e.g. "King(part)"-->"Seattle"
+    if '53' in set(crosswalk['state_fips']):
+        # read file and sheet_name 0 is the state of washington
+        wa = pd.read_excel(nyc_wa_path, sheet_name = 0)
+        # the length of puma code should be 5. Some are not due to data type. 
+        # If it is not 5 digits, add "0" s to make it len()==5
+        wa['PUMA'] = wa['PUMA'].astype('int').astype('str').str.zfill(5)
+        wa = wa[['PUMA','Place']]
+        wa.columns = ['puma_code','new_place']
+        # left join between crosswalk and wa
+        crosswalk = crosswalk.merge(wa, how='left')
+        # meaningful names are in the column 'new_place'. If it is NA in 'new_place', filled by old 'place'
+        crosswalk['new_place'] = crosswalk['new_place'].fillna(crosswalk['place'])
+        # drop 'place'
+        crosswalk = crosswalk.drop('place',axis=1)
+        # rename new_place --> place 
+        crosswalk.rename(columns={'new_place':'place'}, inplace=True)
+
+    #if the file is NYC, replace some 'place' names to meaningful names      
+    if '36' in set(crosswalk['state_fips']):
+        # read file and sheet_name 1 is NYC
+        nyc = pd.read_excel(nyc_wa_path, sheet_name = 1)
+        # add "0" s to make it len()==5
+        nyc['PUMA'] = nyc['PUMA'].astype('int').astype('str').str.zfill(5)
+        nyc = nyc[['PUMA','AreaName']]
+        nyc.columns = ['puma_code','new_place']
+        crosswalk = crosswalk.merge(nyc, how='left')
+        crosswalk['new_place'] = crosswalk['new_place'].fillna(crosswalk['place'])
+        crosswalk = crosswalk.drop('place',axis=1)
+        # rename new_place --> place 
+        crosswalk.rename(columns={'new_place':'place'}, inplace=True)
+    
+    return crosswalk
+
+
+for file in txt_files:
+    df_puma = create_sssplace(puma_crosswalk(file))
+    session.bulk_insert_mappings(PUMA, df_puma.to_dict(orient="records"))
+    session.commit()
+session = DB.sessionmaker()
```

Comment:
Let's put this inside another function, similar to the `data_folder_to_database` function.